doc/Requirements.md

```diff
@@ -0,0 +1,526 @@
+       implements them. Seach for "reference implementation".)
+
+# Introduction
+The PF MUST be a library.
+(Why: To be reused in different components.)
+
+PF instances MUST not share any muttable data.
+(Why: to guarantee that different PF instances will not impact each others.)
+
+# Parameters
+
+A PF MUST be able to handle parameters.
+(Why: because the PF aims to abstract hardware and model it by parameters.)
+
+## Definitions
+
+hardware: system controlled by the PF. Not necessary material system.
+          This term was chosen because:
+ - historically the PF reference implementation was used to abstract hardware
+ - the subsystem term would arguably fit best is already used.
+(FIXME: choose subsystem instead, it is not used in fact.)
+
+## Value
+
+A parameter MUST have a value.
+(Why: because a parameter without value would not abstract any hardware.)
+
+A PF MUST support mutable parameters.
+(Why: To control the underlined hardware.)
+
+A PF MAY support immutable parameters, ie parameters which value is determined on start then read only.
+(Why: To permit hardware read only value reflection.)
+(Note: This is not implemented in the PF reference implementation.)
+
+This value MUST be gettable and settable (except if immutable).
+(Why: A parameter that can not be accessed is of no use.)
+
+A value MUST be convertible to a string.
+(Why: To display a parameter value to the human user for debugging purpose.)
+
+A bijective conversion string <-> value SHOULD exist.
+(Why: To support serialization and deserialization of the parameter values.
+      It is not a MUST because a PF implementation could offer serialization to
+      custom -not string based- format.)
+
+### Data type
+
+#### Definition
+All parameters have a data type.
+A data type designates parameter invariants.
+A data type is the the meaning of the data and the way values of that type can be stored.
+
+
+#### Philosophy
+
+A data type defines the value properties:
+ - memory layout
+ - value constrains
+
+A value type is mostly used to:
+ - pretty display parameter values (not just a as an array of bits)
+ - check for user error when setting it (out of bound, invalid...)
+ - offer a type safe api
+
+#### Requirements
+
+A PF SHOULD support the following types.
+If a type is chosen to be supported, it MUST respect all MUST clause,
+SHOULD respect all SHOULD clause, MAY respect all MAY clause of the type.
+(Why: All type are not necessary to use the PF. For example any parameter could
+      be represented as an array of char (string). But this would not permit to
+      check parameter validity (invariants) as well a pretty displaying the values.)
+
+Implementation MAY add another API to access a parameter value.
+(Why: For example a C++ implementation may give access to a string as an std::string object.)
+
+##### Integers
+PF SHOULD support signed and unsigned integer parameters
+(Why: The reference implementation supports it.)
```

Comment:
not sure why it would not be a MUST?
Maybe you can use "orange" for "why : reference implementation" => all referece implementations to be reviewed, and if relevant, to become MUST I think